### PR TITLE
OP-166 | Download MySQL during build

### LINKED ISSUE #123: OP-166 | Download MySQL during build process
As it's been done for JRE (#121) it'd be better to download MySQL during the build process.

### PR DESCRIPTION
**Pull request by pviotti**
Download MySQL binaries during build, as it is done for JRE.  

Close #123 

**Open issues**
- [x] `Xchang32.exe` is currently used in [`oh.bat`](https://github.com/informatici/openhospital/blob/master/poh-bundle-win/oh.bat) to replace MySQL configuration settings. Such tool is not present in the last version of MySQL (and it should probably be replaced with something like a Powershell command anyway...) [will probably not fix in this PR - there's already too many changes]
 - [x] cache and parallelize downloads
 

**Future improvements** (in other PRs)
 - turn build script in Makefile so that we can skip automatically file that have already been built or downloaded
 - bump poh version (since we changed quite a lot, perhaps it's time to dare a 1.0?)

**Dev notes**
To test locally, set `export OH_VERSION=develop` since the current matching tag across repos might not build successfully. 

for MySQL 8.x these are the URLs:
 - https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.21-linux-glibc2.17-x86_64-minimal.tar.xz
 - https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.21-linux-glibc2.12-i686.tar.xz
 - https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.21-winx64.zip (there is no win32 version)

### CODE REVIEW

File: build_poh.sh
Comment (mwithi) [inline build_poh.sh]: Please add also `mysqldump`. It would be nice also to show the standard output from the OS, e.g.:
```
Command 'mysqldump' not found, but can be installed with:

sudo apt install mysql-client-5.7
sudo apt install mariadb-client-10.1
```

**Review by mwithi — CHANGES_REQUESTED** [review]
1. Please add also `mysqldump` to the requirements. It would be nice also to show the standard output from the OS, e.g.:
```
Command 'mysqldump' not found, but can be installed with:

sudo apt install mysql-client-5.7
sudo apt install mariadb-client-10.1
```

2. When building locally, it needs `rm -r core gui doc` if you want to run more the once. Shall we add this?

3. When dump_error.log contains `mysqldump: Error:` it should stop the build (warnings are allowed) :
```
mysqldump: [Warning] Using a password on the command line interface can be insecure.
mysqldump: Error: 'Access denied; you need (at least one of) the PROCESS privilege(s) for this operation' when trying to dump tablespaces
```

4. It seems the mysqldump error above can be solved with [this](https://dba.stackexchange.com/a/273040) (tested):
```
mysqldump [etc etc]  --no-tablespaces
```

5. at the moment checks are failing because it is fetching v1.10.0 (from `git describe --abbrev=0 --tags`) when tests were relying on the docker mysql instance rather than the in-memory db as it is now, so we could keep running `mvn package` before `docker-compose down`, wdyt?
Tested both: cloning from HEAD with in-memory tests and switching the two commands positions

6. if we `mv` away `*.zip` and `*.gzip` files from root folder, the `checksum` command will fail.

Thanks a lot!!!

**Comment by pviotti** [comment]
1. ack, done (changed mysql to mysqldump, since we don't actually use the mysql binary in the script). About the error message: it depends on the Linux distribution (I use Arch for instance) and it'd be difficult to make if work for different distributions
2. not sure it's the right thing to do. I mean, if you want to start from scratch you can issue `make clean`. And perhaps the best behavior for the script would be to *skip* the git clone if the two folders are already there, instead of removing them (similarly for what happens when `make` finds a target that has already been built - which is why I'm proposing to use more make in #130 )
3. ok, done
4. ok, added
5. ok, moved `docker-compose down` after `mvn package`
6. good catch, should be fixed now

**Comment by mwithi** [comment]
> 1. About the error message: it depends on the Linux distribution (I use Arch for instance) and it'd be difficult to make if work for different distributions

ok

> 2. not sure it's the right thing to do. I mean, if you want to start from scratch you can issue `make clean`. And perhaps the best behavior for the script would be to _skip_ the git clone if the two folders are already there, instead of removing them (similarly for what happens when `make` finds a target that has already been built - which is why I'm proposing to use more make in #130 )

Not sure, if we want to repeat the operation because we know a new commit has landed on one component (and tag moved), it should not _skip_, but _replace_ or _pull_. I've seen #130 but for this PR we should use what we have. Anyway it's now a very big issue, it will prompt an error message and exit, so manually the user can remove.

> Xchang32.exe is currently used in oh.bat to replace MySQL configuration settings. Such tool is not present in the last version of MySQL (and it should probably be replaced with something like a Powershell command anyway...) [will probably not fix in this PR - there's already too many changes]

That command using Xchang32.exe is needed to let the application start, we should fix within this PR as well because otherwise merging it would brake the release. Or we can first do another PR converting oh.bat to Powershell. Or we can provide at least Xchange32.exe to keep the functionality but creating an issue as reminder to replace it.

**Comment by pviotti** [comment]
> Not sure, if we want to repeat the operation because we know a new commit has landed on one component (and tag moved), it should not skip, but replace or pull. 

fair enough, I've added a `rm -rf` before cloning (anyway, if we use tags instead of branches for releases, pulling doesn't help - unless you remove and recreate the tags on different commits..)

> That command using Xchang32.exe is needed to let the application start,

OK, here's a first shot at it:
 - replaced Xchang32 with replace.exe, which takes different arguments - [here's the help page](https://dev.mysql.com/doc/refman/5.7/en/replace-utility.html) (notice it's been removed in v8.0)
 - added `my.ori` since previously it was in `mysql\bin` dir, which has been removed

Notice that:
 - I haven't tested it (I don't have a Windows machine handy for testing it now..)
 - I have no idea what those [`Xchang32 "^x5c" "^x2f"`](https://github.com/informatici/openhospital/blob/819b7b226a199863a72888416a1e4266df5acbf2/poh-bundle-win/oh.bat#L37) do and why they are there

Additional changes (since this PR turned out to be about general improvements on the script..):
 - increased parallelism of Maven build
 - use `--no-clobber` option in wget to avoid downloading files if they're already there (notice that if the files are corrupted or empty it does not download them either)
 - add `clean-download` target to makefile
 - get version to build optionally from environment variable, e.g., `export OH_VERSION=develop; make` will build from develop branches
 - etc, minor clean ups

**Comment by mwithi** [comment]
> anyway, if we use tags instead of branches for releases, pulling doesn't help - unless you remove and recreate the tags on different commits.

Yeah, I meant that, it happens a lot of times during the release process (moving tags). Should we remove also previous created folders?
`replace ./poh-win32-0.0.6-core-v1.10.0/mysql-5.7.30-win32/bin/echo.exe? [y]es, [n]o, [A]ll, [N]one, [r]ename: A`

> I haven't tested it (I don't have a Windows machine handy for testing it now..)

Tested and the build works like a sharm! Congrats! 😀
Not so smooth the application startup, see list below

> I have no idea what those Xchang32 "^x5c" "^x2f" do and why they are there

Probably converting Unix-LF to Windows-CRLF

> get version to build optionally from environment variable, e.g., export OH_VERSION=develop; make will build from develop branches

Very usefull!!! Thanks a lot!!!

### Problems on startup (hope solved with https://github.com/informatici/openhospital/pull/129/commits/c69c004b4fc00b8a32a392870d23dd0d5fbf7921)
- [x] It gets stuck when using `replace.exe`: removed quotes and commented out "^x5c" "^x2f" replacement
- [x] Also, `my.ori` should be searched in `cd /d %OH_PATH%`
- [x] If the command   `%OH_PATH%mysql-5.7.30-win32\bin\mysql -u root --port=%freePort% oh < "%OH_PATH%database.sql"` fails, it should not delete `database.sql` file
- [x] To download the mysql-package is not enough, `--initialize` should be used first to generate the `data` folder
- [x] `my.ori` needs paths adjusting to the new mysql folder, anyway since v5.7 there is an [Option File Processing Order](https://dev.mysql.com/doc/refman/5.7/en/option-files.html#option-file-order) and [Inclusion](https://dev.mysql.com/doc/refman/5.7/en/option-files.html#option-file-inclusions) syntax, so the easiest solution I found was to move `my.cnf` to the `mysql/bin` folder
- [x] mysqldump should not need anymore `--compatible=mysql40` option, I've removed so we can test it
- [x] the 'oh' database is missing, so we need to create it and also the users: adding same commands as in `step_00_create_user.sql` to the initialization step
- [x] probably some things apply to 'oh.sh' too

**Review by mwithi — APPROVED** [review]

**Comment by pviotti** [comment]
I confirm it works on Windows (though I had some issue with too long path when starting oh.bat so I had to move the folder somewhere else - probably it'll get fixed once we migrate to powershell...).  
I also removed the build dependency from mysqldump, and used the one in the running container instead (this is also because I got [this error](https://serverfault.com/questions/912162/mysqldump-throws-unknown-table-column-statistics-in-information-schema-1109), which is due to some flag in v8.0 of the client that I had installed locally).
Let's check also Linux builds and then this can be merged.

**Comment by mwithi** [comment]
> I confirm it works on Windows (though I had some issue with too long path when starting oh.bat so I had to move the folder somewhere else - probably it'll get fixed once we migrate to powershell...).

Yes, known problem (OP-179) and yes, PS 👍 

> I also removed the build dependency from mysqldump, and used the one in the running container instead (this is also because I got this error, which is due to some flag in v8.0 of the client that I had installed locally).

Nice shot!

> Let's check also Linux builds and then this can be merged.

Ok, I will post here the results!

**Comment by mwithi** [comment]
Running on linux32 (Ubuntu16.10.1 in VM):

- `bash: ./oh.sh: /bin/bash^M: bad interpreter: No such file or directory`
- changed with `dos2unix` utility, it hangs at the DB initialization, I think for these same reason as in Windows (it needs MySQL Server initialization)
- with `mysqld --initialize-insecure` it shows: 
```
./mysqld: error while loading shared libraries: libaio.so.1: cannot open shared object file: No such file or directory
``` 
it seems [required](https://dev.mysql.com/doc/refman/8.0/en/binary-installation.html)
- after installing libs with `sudo apt-get install libaio1 libaio-dev` it shows:
```
mysqld: Can't create directory '/usr/local/mysql/data/'
```

Maybe the procedure for standalone MySQL server are slight different on Linux

**Comment by mwithi** [comment]
Note: the downloaded mysql-5.7.30-linux-glibc2.12-i686 is 2GB....

In Linux, another option is to use Docker, that should be available by default in all 64bit Linux systems, at least... we already have the dockerfile and docker-compose. Migrating the image from the container started during the build is quite easy, less easy is the migration of the volume in which the DB is built, probably would be better to recreate the DB on the user machine (given the image)

Note: Dockerfile uses mysql:5.6 at the moment
Note2: the problem remains for 32bit linux clients

_Proposed_ Actions: 
- to bump image in Docker from mysql:5.6 to mysql:5.7 in core
- to provide only portables versions (embedded JVM) including scripts, and so avoid the "full" version
- to provide only 64bit support for Linux and to introduce Docker for the db (bound to a subfolder?)

Wdyt?

**EDIT**: to download Docker for users could be overkilling, so we discourage this option at the moment

**Comment by mwithi** [comment]
Size is quite double compared to previous release... can we do more?

File: poh-bundle-linux-x32/oh.sh
Comment (pviotti) [inline poh-bundle-linux-x32/oh.sh]: @mwithi duplicate command

File: poh-bundle-linux-x64/oh.sh
Comment (pviotti) [inline poh-bundle-linux-x64/oh.sh]: @mwithi duplicate command

**Comment by pviotti** [comment]
> Size is quite double compared to previous release... can we do more?

That's the price to pay when you want to create a "portable" distribution out of a full fledged RDBMS... :)
We may do better, or maybe not. After 42 commits I'd be inclined to merge this PR if things work, then we can reason on what can be cut (if any).

File: poh-bundle-linux-x64/oh.sh
Comment (mwithi) [reply poh-bundle-linux-x64/oh.sh]: why duplicate? it is a different path... my.cnf expects that the path for the log file exists

File: poh-bundle-linux-x32/oh.sh
Comment (mwithi) [reply poh-bundle-linux-x32/oh.sh]: it's a different path...

Comment (pviotti) [reply poh-bundle-linux-x32/oh.sh]: yup sorry, my bad

**Pull request merged**
